apps/examples/simple_file_transfer/sft_main.c

```diff
@@ -0,0 +1,316 @@
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/stat.h>
+#include <net/if.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/select.h>
+#include <sys/socket.h>
+#include <pthread.h>
+
+#define SFT_PORT 5555
+#define BUF_SIZE 512
+char g_buf[BUF_SIZE] = {0,};
+char g_name_buf[128];
+
+static int send_data(int sd, char *buf, int buf_size)
+{
+	printf("-->send data %d\n", buf_size);
+	int sent = 0;
+	while (sent < buf_size) {
+		int ret = send(sd, buf + sent, buf_size - sent, 0);
+		if (ret <= 0) {
+			if (ret == 0) {
+				printf("connection closed\n");
+				return -1;
+			}
+			if (errno == EWOULDBLOCK) {
+				continue;
+			}
+		}
+		sent += ret;
+	}
+	return 0;
+}
+
+static int read_data(int sd, char *buf, int buf_size, int *readed)
+{
+	int read_size = 0;
+	while (read_size < buf_size) {
+		int nbytes = recv(sd, buf + read_size, buf_size - read_size, 0);
+		if (nbytes == 0) {
+			printf("connection closed\n");
+			return -1;
+		} else if (nbytes < 0) {
+			printf("unknown error(%d)\n", errno);
+			return -1;
+		}
+		read_size += nbytes;
+		*readed = read_size;
+	}
+	return 0;
+}
+
+static int write_data(FILE *fp, char *buf, int buf_size)
+{
+	int written = 0;
+	while (written < buf_size) {
+		int res = fwrite(buf + written, sizeof(char), buf_size - written, fp);
+		if (res < 0) {
+			printf("file write fail %d\n", errno);
+			return -1;
+		}
+		written += res;
+	}
+	return 0;
+}
+
+static int get_file_size(FILE *fp)
+{
+	fseek(fp, 0, SEEK_END);
+	int file_size = ftell(fp);
+
+	printf("file_size (%d)\n", file_size);
+	fseek(fp, 0, SEEK_SET);
+
+	return file_size;
+}
+
+static void send_file(int connfd)
+{
+	int readed = 0;
+	int file_name_size = 0;
+	int ret = read_data(connfd, (char *)&file_name_size, sizeof(int), &readed);
+	if (ret < 0) {
+		printf("readed %d %d\n", readed, __LINE__);
+		return;
+	}
+	file_name_size = ntohl(file_name_size);
+	printf("file name length(%d) %d\n", file_name_size, readed);
+
+	ret = read_data(connfd, g_name_buf, file_name_size, &readed);
+	if (ret < 0) {
+		printf("fail %d\n", __LINE__);
+		return;
+	}
+	g_name_buf[file_name_size] = 0;
+	printf("file name: %s\n", g_name_buf);
+
+	FILE *fp = fopen(g_name_buf, "r");
+	if (!fp) {
+		printf("fail %d\n", __LINE__);
+		return;
+	}
+
+	//get file size
+	int file_size = get_file_size(fp);
+	printf("send file size %d\n", file_size);
+	int net_file_size = htonl(file_size);
+	ret = send_data(connfd, (void *)&net_file_size, sizeof(int));
+	if (ret < 0) {
+		printf("error %d\n", __LINE__);
+		return;
```

Comment:
Is it okay not to close the `connfd` and `fp`?

Reply:
connfd is closed at tcp_server()